condition_resource_test.go

```diff
@@ -0,0 +1,48 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ladon
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestResourceMatch(t *testing.T) {
+	for _, c := range []struct {
+		matches   string
+		delimiter string
+		resource  string
+		value     string
+		pass      bool
+	}{
+		{delimiter: ":", value: "GroupId:2", resource: "resources:sensor-data:GroupId:2:deviceIds:a9b576e8-7419-4eed-a010-7f68ec0ff588", pass: true},
+		{delimiter: "_", value: "GroupId_2", resource: "resources_sensordata_GroupId_2_deviceIds_a9b576e8-7419-4eed-a010-7f68ec0ff588", pass: true},
+		{delimiter: "_", value: "deviceIds_a9b", resource: "resources_sensordata_GroupId_2_deviceIds_a9b", pass: true},
+		{delimiter: "_", value: "deviceIds_a9b", resource: "resources_sensordata_GroupId_2_deviceIds_a9b6", pass: false},
+		{delimiter: ":", value: "GroupId:2", resource: "resources:sensor-data:GroupId:2:deviceIds:*", pass: true},
+		{delimiter: ":", value: "GroupId:2", resource: "abc", pass: false},
+		{delimiter: ":", value: "", resource: "abc", pass: false},
+		{delimiter: "", value: "test", resource: "abc", pass: false},
+	} {
```

Comment:
The tests are not really expressive at the moment, I'm not 100% sure what should fail and what shouldn't. For example, should this fail?

* {delimiter: ":", value: "GroupId:2", resource: "resources:sensor-data:GroupId:223:deviceIds:a9b576e8-7419-4eed-a010-7f68ec0ff588"},
* {delimiter: ":", value: "GroupId:2", resource: "resources:sensor-data:GroupId:2"},

If so, please add more tests to express this :)

Reply:
I am afraid (but not certain) that the regex could be a lot slower in comparison to the contains, since we need to do this resource check on a lot of requests. Did you take that into consideration for condition string match?

https://awmanoj.github.io/tech/2016/09/08/on-slow-regular-expressions-golang/

Reply:
210µs vs 40µs on average for the test cases... i.e. more than 5 times slower...

Reply:
Regex is slower (it has O(n) in Go) than simple string comparison.

What I fear about this PR is that it is aimed quite specific at one use case (could you please take the time to answer the question regarding what fails and wont so I get a better understanding of this) which could be generalized with regular expressions.

If you have fewer than 10000 requests per second (~1B requests per day) I wouldn't worry about this too much ;)

Reply:
Yes sure will, thanks for your answer, good points. regex should be used for making this functionality more generic and usable...